### DISCUSSION TITLE
IEnumerable.GetEnumerator with default implementation

### DISCUSSION THREAD

**Discussion by dellamonica** [dotnet/csharplang#7569] [General]
The generic IEnumerable<T> interface inherits from the (legacy) non-generic IEnumerable. They both have GetEnumerator methods which differ on the return type by the generic vs non generic versions of IEnumerator. Anytime someone needs to implement IEnumerable<T> (and I imagine essentially no-one works with the non-generic collections at this time) it is also necessary to implement the non-generic version for backwards compatibility polluting the code with outdated concepts.

The question here is: couldn't the BCL be updated so that the non-generic version has a default implementation which just calls the generic version?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/runtime/issues/29513

**Reply by dellamonica** [reply]
I'm not really convinced the risk mentioned in that thread is even real... but I understand being cautious.

IMO, C# has some baggage from its initial years that sometimes surfaces in newer code, so finding ways of hiding that cruft with modern features is welcomed (of course, provided it doesn't cause more harm than good).

**Reply by CyrusNajmabadi** [reply]
> I'm not really convinced the risk mentioned in that thread is even real...

The risk is very real.  Especially for what is arguably one of the most critical types in all of .net :)

We're talking about trillions of lines of code at this point that depends on this in some way, and which has been built a certain way for 25 years now.  Destabilizing that in any way is taken very seriously.  :)

**Reply by dellamonica** [reply]
The only way to be affected is to explicitly recompile using a newer version of the framework, and to have dependencies on libraries that do weird stuff with their interfaces. So I think the vast majority of those trillions of lines would be be just fine. The same argument can be made for most changes, so it's always a cost-benefit analysis.

It seems clear to me you are not gonna make a move on this, so feel free to close the discussion, however, I really think the language team should have a strategy to get rid of the legacy cruft in the language. The non-generic collections is certainly one of the warts, but there are more, and you know. Perhaps there could be a compilation mode for "fresh" codebases in which the legacy stuff can be entirely stripped away.

**Reply by HaloFour** [reply]
Those aren't language concerns, the language works with whatever types and APIs are provided by the runtime and referenced libraries.  It'd be up to the runtime team to decide to use default interface methods, or to remove existing APIs.  But doing so is not to be taken lightly given that breaking changes can result in long-lived and very difficult to reconcile schisms in the ecosystem.  The .NET community is still recovering from the switch to .NET Core from .NET Framework.

**Reply by CyrusNajmabadi** [reply]
> It seems clear to me you are not gonna make a move on this, so feel free to close the discussion

We're not going to move on this in the short term.  But the discussion is relevant and valuable for the long term.  For context, we're already 25 years old.  We have to think and plan over those sorts of timescales :)

> however, I really think the language team should have a strategy 

This is not part of the language.  You can def send that feedback to the runtime though :)

> to get rid of the legacy cruft in the language.

One person's 'legacy cruft' is another person's invaluable base that they depend on.  We have ample experiences with breaking changes and the benefits/costs around there.  Right now this isn't crossing the necessary hurdle to make it worthwhile.  That may change in the future.

**Reply by DavidArno** [reply]
> Those aren't language concerns, the language works with whatever types and APIs are provided by the runtime and referenced libraries. It'd be up to the runtime team to decide to use default interface methods, or to remove existing APIs. But doing so is not to be taken lightly given that breaking changes can result in long-lived and very difficult to reconcile schisms in the ecosystem. The .NET community is still recovering from the switch to .NET Core from .NET Framework.

The fact that "_the .NET community is still recovering from the switch to .NET Core from .NET Framework_" calls out the fallacy of the making changes "_is not to be taken lightly_". Any change will cause problems, and will result in pushback from those that dislike change. So there is zero point in pandering to such people. They will moan no matter what you do. .NET Core was a missed opportunity to have a proper clear out of the technical debt crud that litters the BCL. If you are going to break something, make the effort needed to fix things worthwhile. So we've all had to go through the pain of migrating to .NET Core, but haven't benefitted from eg watching the non-generic collection types fall by the wayside as we did so. 

MS seem determined to never learn from constantly repeating the "_we are going to make our lives, and our customer's lives, harder by not making sensible breaking changes_" mistake. 🤷

**Reply by HaloFour** [reply]
> .NET Core was a missed opportunity to have a proper clear out of the technical debt crud that litters the BCL.

.NET Core went to some effort to do just that.  Some remain gone, some had to be put back.  It's compromise, and always will be, and it _still_ resulted in a schism.  But the goal isn't to make a hard break so that only new projects can adopt, the goal is to provide a migration path to existing users.

That assumes that the .NET runtime team thinks those APIs are legacy cruft, and they probably don't.

**Reply by CyrusNajmabadi** [reply]
> MS seem determined to never learn

We learned first-hand actually trying this.  The community *at large* did not actually behave as you said they would. As such, we revised and compromised.  That upset a very tiny part of the user base. And importantly, while it upset them, it did not block adoption. 

So given the choice of making that group happy, but having severe issues with adoption, or making the majority happy, and getting great adoption, I think we made the right choice. :-)

What we've found is that there are a tiny handful of hardliners. But that tend to be unhappy no matter what. And at the end of the day, they'll still come along. So better to just be pragmatic and accommodating than try to please that group at the expense of the ecosystem as a whole.

**Reply by dellamonica** [reply]
I cannot imagine why would the non-generic collection types be anything but cruft, please illuminate me? One can always use `object` as the generic type and I think the migration path to replace those usages is straightforward.

**Reply by CyrusNajmabadi** [reply]
Lots of projects reference code that is not being updated. This is common. Removing the non-generics entirely breaks this. It's a hard fork off, bifurcating things strongly, instead of allowing the vast majority to have nearly all the scenarios work.

**Reply by HaloFour** [reply]
Removing APIs for functionality that no longer exists in the runtime is one thing, especially when the only choice is to have it throw an exception.  These collections may be largely supplanted, but they're also benign.  It would require more work to remove them and cause more pain for those trying to move forward from .NET Framework.  The issue is viral, too, based on your dependencies.

**Reply by dellamonica** [reply]
I'm not disputing any of this, just insisting that **this stuff is indeed cruft** or "tech debt" as @DavidArno mentioned... a hard fork sometimes is necessary, but maybe a clean up could be accomplished without breaking. 

I'm still thinking about what the runtime/BCL could do as a more general strategy, but perhaps moving the legacy types into a separate *package* that would be included by projects using old code that is being migrated to newer frameworks. The runtime support would be needed in order to "compose" the BCL based on which packages are selected. Clearly the BCL was not designed to be modular, so I understand this is a tall order, but the benefits are also for the continued evolution of the framework, since what is shiny today will be obsolete some time in the future.

**Reply by CyrusNajmabadi** [reply]
> I'm still thinking about what the runtime/BCL could do as a more general strategy, but perhaps moving the legacy types into a separate package

But how would this work here with IEnumerable?  `IEnumerable<T>` inherits from it.  

> Clearly the BCL was not designed to be modula

The BCL already has many modules, and already allows thigns like type-forwarding to move things into different packages.  However, that's not really helpful because people don't want a fine grained set of packages they have to then navigate.  They want a standard set they know that their code will continue to work well on.  Hence the different standard sets that just pull everything in in ways that try to keep high compat.

**Reply by dellamonica** [reply]
> But how would this work here with IEnumerable? IEnumerable<T> inherits from it.

On a high level, both "legacy" and "modern" packages would both define the interface `IEnumerable<T>` but on the "modern" package `IEnumerable` would not be defined and thus `IE<T>` would not inherit from it.

Namespaces and shims could be used to bridge the gap when upgrading. So maybe if you recompile a project trying to use a modern version but with some dependency that relies on the legacy types, those dependencies would have a shim that replaces the namespace of the legacy types to something like `System.Legacy.NonGeneric` where `IEnumerable` and friends are defined. 

Not sure how well this would work but you could define an interface `S.L.NG.IEnumerable<T>` that inherits both from the modern `IEnumerable<T>` and the legacy `IEnumerable`, which means that any type implementing the "legacy" `IEnumerable<T>` in the dependency would be compatible with the client's code that uses the modern package.

> people don't want a fine grained set of packages they have to then navigate.

Agree, but you don't need a toggle for every little bit, you could have simply a "modern" and a "legacy" version for each dotnet release, and what is removed from the legacy version can evolve with the releases. Many people won't mind keeping all the legacy stuff around, so they won't have to worry about anything.

**Reply by CyrusNajmabadi** [reply]
> Namespaces and shims could be used to bridge the gap when upgrading. 

Sicne all those work has to be done.  It just means that things are more complex anyways, and potentially even confusing and error prone for users.  Just having the type exist in its current state and continue working means people don't have to run into this, the costs are zero, and the risk are as well.

**Reply by dellamonica** [reply]
> Sicne all those work has to be done. It just means that things are more complex anyways, and potentially even confusing and error prone for users. Just having the type exist in its current state and continue working means people don't have to run into this, the costs are zero, and the risk are as well.

Again, not arguing that this is worthy it for this specific case, you may as well be entirely correct in the cost-benefit analysis. However, not having any decent strategy for deprecating code in the BCL is IMO a flaw of dotnet. As you said, it's 25 years old and it will only get older. I'll concede that the dotnet community efforts to make trimming better in the recent years at least helps to remove the old stuff from published binaries.

**Reply by CyrusNajmabadi** [reply]
We do deprecate things.  But generally with a cost/benefit/risk assessment that warrants it being worthwhile.  I think where there is disagreement here is on the enormity of the risk.  You're talking about what could literally be described as one of the top 10 types in all of .net.  It would be akin to changing int32, or string.  etc.  Such a change is going to need very large benefit, or just huge assurance that it's going to be safe to make it worthwhile.  Absent that, it's not going to happen as there is just likely to be real pain given the numbers we're talking here, and only very marginal benefit.

**Reply by dellamonica** [reply]
For good or bad, dotnet is very conservative with respect to deprecation.

I'm not sure I'll follow up with the runtime team as the only feasible plan I could come up with for general deprecation of such fundamental types as IEnumerable does seem very costly in terms of tooling and friction with the user base.

Thank you all for this nice discussion.

**Comment by dellamonica** [comment]
> > It seems clear to me you are not gonna make a move on this, so feel free to close the discussion
> 
> We're not going to move on this in the short term. But the discussion is relevant and valuable for the long term. For context, we're already 25 years old. We have to think and plan over those sorts of timescales :)
> 

Yeah, I know, I've been using C# for almost that long.

> > however, I really think the language team should have a strategy
> 
> This is not part of the language. You can def send that feedback to the runtime though :)

I think it needs some support from the language, even if in the form of analyzers, but sure, the runtime needs to be worked on, not sure I have a concrete model for the more general idea of removing obsoleted parts of the BCL from "fresh" codebases. Let me think and write about it and I will link it up here.

> > to get rid of the legacy cruft in the language.
> 
> One person's 'legacy cruft' is another person's invaluable base that they depend on. We have ample experiences with breaking changes and the benefits/costs around there. Right now this isn't crossing the necessary hurdle to make it worthwhile. That may change in the future.

Most people would agree that the non-generic collections are legacy cruft, and for those that depend on it still, they are surely not targeting .net 8.
